node/lib/util/rm.js

```diff
@@ -0,0 +1,482 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert       = require("chai").assert;
+const binarySearch = require("binary-search");
+const co           = require("co");
+const fs           = require("fs-promise");
+const path         = require("path");
+const NodeGit      = require("nodegit");
+
+const CloseUtil           = require("./close_util");
+const StatusUtil          = require("./status_util");
+const SubmoduleUtil       = require("./submodule_util");
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const TextUtil            = require("./text_util");
+const UserError           = require("./user_error");
+
+const FILEMODE = NodeGit.TreeEntry.FILEMODE;
+
+const compareIndexEntryToPath = function(indexEntry, path) {
+    return TextUtil.strcmp(indexEntry.path, path);
+};
+
+const needToRequestRecursive = function(path) {
+    throw new UserError(
+        `not removing '${path}' recursively without -r`);
+};
+
+/**
+ * helper for getRemovalSpecs
+ * @param first The first index to look at
+ */
+/**
+ * There are four possible things that a user might be trying to
+ * remove:
+ *
+ * 1. A file inside a submodule
+ * 2. A subdirectory inside a submodule
+ * 3. A submodule or file within the meta repo
+ * 4. A directory containing multiple such things
+ *
+ * In order to handle cases 2 and 4, we will delegate to nodegit2's
+ * index.removeAll.  We need to ensure that there is at least one
+ * object in the (meta or sub) index that matches the user's request;
+ * if there is not, we throw an exception.
+ *
+ * @visibleForTesting
+ *
+ * @async
+ * @param {String} pathToRemove Normalized to the root of the meta
+ * repo, and no trailing slashes.  In the event that the user desires
+ * to remove everything in the repo, the pathspec "." must be
+ * translated to "", which is illegal from the git CLI's perspective,
+ * but which is legit internally.
+ *
+ * @param {Boolean} recursive If false, and the path to remove is a directory,
+ * give the "can't remove a dir without -r" message
+ *
+ * @return {Object}
+ * @return {Boolean} return.isRootLevel -- distinguishes cases 1/2 from 3/4
+ * @return {String}  return.pathFromRoot to the submodule
+ * @return {String}  return.pathInSubmodule
+ * @return {NodeGit.Repo}  return.submoduleRepo
+ */
+exports.getRemovalSpec = co.wrap(function *(repo, index, pathToRemove,
```

Comment:
As this started to soak in today, I found it very unfortunate that this was so much more complicated than `add`, which needs to do essentially the same logic and is only 77 lines of code (a little more than 1/10th of this). Of course, `add` has the advantage that it is looking for only *changed* files, and so can take utilize the `pathspec` in, e.g., `NodeGit.Diff.treeToWorkdir`.

However, I think with a little tweaking you could do the same thing as `add`. It might require some refactoring of `StatusUtil.getRepoStatus` and/or `DiffUtil.getRepoStatus`, depending on how you want to slice it, but at the crux of it you'd be operating based on a `Diff` constructed by running `NodeGit.Diff.treeToIndex` against an *empty* tree. The big advantage here is that you can let libgit2 do the nasty work of handling the `pathspec`. You could probably save a lot of effort even if you used `NodeGit.Diff.treeToIndex` directly rather than attempting to factor `DiffUtil` or `StatusUtil`.

Up to you, but I'll let you think on it before giving more feedback.